// @jxpeng98/markdown-it-citation@1.2.1/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = citationPlugin;
// import StateInline from 'markdown-it/lib/rules_inline/state_inline';
const fs_1 = __importDefault(require("fs"));
const js_yaml_1 = __importDefault(require("js-yaml"));
// Use require for citation-js to avoid ESM/CJS interop issues
const Cite = require('citation-js');
// Load CSL plugin for citation-js to support different styles
require('@citation-js/plugin-csl');
const CITATION_RE = /(!?)@([a-zA-Z0-9_:-]+)/;
const FRONTMATTER_RE = /^---\n([\s\S]+?)\n---/;
const BIBLIOGRAPHY_MARKER_RE = /^\[\[bibliography\]\]$/i;
function autoLinkUrlsAndDois(html) {
    // Auto-link plain HTTP(S) URLs
    html = html.replace(/(https?:\/\/[^\s<>"']+)/g, (match) => {
        return `<a href="${match}" target="_blank" rel="noopener noreferrer">${match}</a>`;
    });
    // Auto-link bare DOIs that appear as "doi: 10.xxxx/..." or "DOI: 10.xxxx/..."
    html = html.replace(/(doi:\s*)(10\.\d{4,9}\/[^\s<>"']+)/gi, (full, prefix, doi) => {
        const url = `https://doi.org/${doi}`;
        return `${prefix}<a href="${url}" target="_blank" rel="noopener noreferrer">${doi}</a>`;
    });
    return html;
}
function citationPlugin(md, options = {}) {
    const bibFile = options.bibFile || 'public/references.bib';
    const style = options.style || 'apa';
    const showNum = options.showNum || false;
    const autoGenerate = options.autoGenerate ?? false; // Default to false (manual trigger)
    let bibContent = '';
    try {
        if (fs_1.default.existsSync(bibFile)) {
            bibContent = fs_1.default.readFileSync(bibFile, 'utf-8');
        }
        else {
            console.warn(`[markdown-it-citation] Bibliography file not found: ${bibFile}`);
        }
    }
    catch (e) {
        console.error(`[markdown-it-citation] Error reading bibliography file:`, e);
    }
    // Initialize Cite instance with bibliography content
    // Handle empty content gracefully
    let cite;
    try {
        cite = new Cite(bibContent || '');
    }
    catch (e) {
        console.error('[markdown-it-citation] Failed to parse bibliography:', e);
        cite = new Cite();
    }
    // Block rule to parse [[bibliography]] marker
    md.block.ruler.before('paragraph', 'bibliography_marker', (state, startLine, endLine, silent) => {
        const pos = state.bMarks[startLine] + state.tShift[startLine];
        const max = state.eMarks[startLine];
        const lineText = state.src.slice(pos, max).trim();
        if (!BIBLIOGRAPHY_MARKER_RE.test(lineText)) {
            return false;
        }
        if (silent) {
            return true;
        }
        const token = state.push('bibliography_placeholder', '', 0);
        token.map = [startLine, startLine + 1];
        token.content = '';
        state.line = startLine + 1;
        return true;
    });
    // Inline rule to parse @citekey in text
    md.inline.ruler.push('citation', (state, silent) => {
        const src = state.src.slice(state.pos);
        const match = src.match(CITATION_RE);
        if (!match || match.index !== 0) {
            return false;
        }
        if (silent) {
            return true;
        }
        const isNarrative = match[1] === '!';
        const key = match[2];
        const token = state.push('citation', 'span', 0);
        token.attrs = [
            ['data-key', key],
            ['data-narrative', isNarrative ? 'true' : 'false']
        ];
        token.content = match[0];
        token.info = key;
        state.pos += match[0].length;
        return true;
    });
    // Renderer for citation token
    md.renderer.rules.citation = (tokens, idx) => {
        const key = tokens[idx].attrs?.[0][1];
        const isNarrative = tokens[idx].attrs?.[1][1] === 'true';
        if (!key)
            return '';
        try {
            let citationText = '';
            if (isNarrative) {
                // Narrative citation: Author (Year)
                const entry = cite.data.find((e) => e.id === key);
                if (!entry)
                    throw new Error(`Entry ${key} not found`);
                let author = 'Unknown';
                if (entry.author && entry.author.length > 0) {
                    if (entry.author.length === 1) {
                        author = entry.author[0].family;
                    }
                    else if (entry.author.length === 2) {
                        author = `${entry.author[0].family} & ${entry.author[1].family}`;
                    }
                    else {
                        author = `${entry.author[0].family} et al.`;
                    }
                }
                else if (entry.editor && entry.editor.length > 0) {
                    author = entry.editor[0].family; // Fallback to editor
                }
                const year = entry.issued?.['date-parts']?.[0]?.[0] || 'n.d.';
                citationText = `<span class="citation narrative" data-key="${key}">${author} (${year})</span>`;
            }
            else {
                // Standard parenthetical citation: (Author, Year) or [1]
                const entry = cite.format('citation', {
                    entry: key,
                    template: style,
                    lang: 'en-US'
                });
                citationText = `<span class="citation" data-key="${key}">${entry}</span>`;
            }
            // Wrap in anchor tag for linking
            return `<a href="#ref-${key}" class="citation-link">${citationText}</a>`;
        }
        catch (e) {
            return `<span class="citation-error" data-key="${key}">[?${key}]</span>`;
        }
    };
    // Core rule to append bibliography
    md.core.ruler.push('citation_bibliography', (state) => {
        const usedKeys = new Set();
        // 1. Collect keys from in-text citations
        state.tokens.forEach(token => {
            if (token.type === 'inline' && token.children) {
                token.children.forEach(child => {
                    if (child.type === 'citation') {
                        const key = child.attrs?.[0][1];
                        if (key)
                            usedKeys.add(key);
                    }
                });
            }
        });
        // 2. Collect keys from Frontmatter
        let frontmatter = (state.env && state.env.frontmatter);
        if (!frontmatter) {
            const match = state.src.match(FRONTMATTER_RE);
            if (match) {
                try {
                    frontmatter = js_yaml_1.default.load(match[1]);
                }
                catch (e) {
                    // Ignore parsing errors
                }
            }
        }
        if (frontmatter) {
            const possibleKeys = ['citations', 'references', 'bib'];
            for (const key of possibleKeys) {
                if (frontmatter[key]) {
                    const val = frontmatter[key];
                    if (Array.isArray(val)) {
                        val.forEach((k) => {
                            const cleanKey = k.startsWith('@') ? k.slice(1) : k;
                            if (cleanKey.startsWith('!')) {
                                usedKeys.add(cleanKey.slice(1));
                            }
                            else {
                                usedKeys.add(cleanKey);
                            }
                        });
                    }
                    else if (typeof val === 'string') {
                        const keys = val.split(/[\s,]+/);
                        keys.forEach(k => {
                            let cleanKey = k.startsWith('@') ? k.slice(1) : k;
                            if (cleanKey.startsWith('!'))
                                cleanKey = cleanKey.slice(1);
                            if (cleanKey)
                                usedKeys.add(cleanKey);
                        });
                    }
                }
            }
        }
        // Helper function to generate bibliography HTML
        const generateBibliographyHtml = () => {
            try {
                const data = cite.data.filter((entry) => usedKeys.has(entry.id));
                if (data.length === 0)
                    return '';
                const subCite = new Cite(data);
                let bibOutput = subCite.format('bibliography', {
                    format: 'html',
                    template: style,
                    lang: 'en-US'
                });
                // Add IDs to entries for linking
                bibOutput = bibOutput.replace(/data-csl-entry-id="([^"]+)"/g, (match, key) => {
                    return `id="ref-${key}" ${match}`;
                });
                // Ensure URLs and DOIs inside bibliography entries are clickable links
                bibOutput = autoLinkUrlsAndDois(bibOutput);
                const containerClass = showNum ? 'bibliography bibliography-numbered' : 'bibliography';
                const titleHtml = options.bibTitle ? `<h2 class="bibliography-title">${options.bibTitle}</h2>\n` : '';
                return `<div class="${containerClass}">\n${titleHtml}${bibOutput}\n</div>`;
            }
            catch (e) {
                console.error("[markdown-it-citation] Error generating bibliography:", e);
                return '';
            }
        };
        // 3. Find [[bibliography]] placeholder and replace it, or auto-generate at end
        let hasPlaceholder = false;
        for (let i = 0; i < state.tokens.length; i++) {
            if (state.tokens[i].type === 'bibliography_placeholder') {
                hasPlaceholder = true;
                const bibHtml = generateBibliographyHtml();
                if (bibHtml) {
                    state.tokens[i].type = 'html_block';
                    state.tokens[i].content = bibHtml;
                }
                else {
                    // Remove the placeholder if no bibliography to generate
                    state.tokens.splice(i, 1);
                    i--;
                }
            }
        }
        // 4. Auto-generate bibliography at end if enabled and no placeholder found
        if (!hasPlaceholder && autoGenerate) {
            const bibHtml = generateBibliographyHtml();
            if (bibHtml) {
                const token = new state.Token('html_block', '', 0);
                token.content = bibHtml;
                state.tokens.push(token);
            }
        }
    });
}
